File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/model/SmartMockHeaders.java
package com.crescentflare.smartmock.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Smart mock library model: a set of headers
 */
public class SmartMockHeaders
{
    /**
     * Members
     */

    private Map<String, List<String>> values = new HashMap<>();


    /**
     * Initialization
     */

    private SmartMockHeaders()
    {
        // Private constructor, use factory methods to create an instance
    }

    public static SmartMockHeaders create(Map<String, List<String>> headers)
    {
        SmartMockHeaders result = new SmartMockHeaders();
        if (headers != null)
        {
            result.values = headers;
        }
        return result;
    }

    public static SmartMockHeaders createFromFlattenedMap(Map<String, String> headers)
    {
        SmartMockHeaders result = new SmartMockHeaders();
        if (headers != null)
        {
            for (String key : headers.keySet())
            {
                List<String> list = new ArrayList<>();
                list.add(headers.get(key));
                result.values.put(key, list);
            }
        }
        return result;
    }


    /**
     * Access headers
     */

    public Map<String, List<String>> getHeaderMap()
    {
        return values;
    }

    public void overwriteHeaders(SmartMockHeaders headers)
    {
        for (String key : headers.getHeaderMap().keySet())
        {
            setHeader(key, headers.getHeaderValue(key));
        }
    }

    public String getHeaderValue(String key)
    {
        for (String checkKey : values.keySet())
        {
            if (checkKey.equalsIgnoreCase(key))
            {
                String result = "";
                for (String value : values.get(checkKey))
                {
                    if (result.length() > 0)
                    {
                        result += "; ";
                    }
                    result += value;
                }
                return result;
            }
        }
        return null;
    }

    public void setHeader(String key, String value)
    {
        List<String> list = new ArrayList<>();
        list.add(value);
        removeHeader(key);
        values.put(key, list);
    }

    public void addHeader(String key, String value)
    {
        for (String checkKey : values.keySet())
        {
            if (checkKey.equalsIgnoreCase(key))
            {
                values.get(checkKey).add(value);
                return;
            }
        }
        setHeader(key, value);
    }

    public void removeHeader(String key)
    {
        for (String checkKey : values.keySet())
        {
            if (checkKey.equalsIgnoreCase(key))
            {
                values.remove(checkKey);
                return;
            }
        }
    }
}

File: MockLibIOS/SmartMockLib/Classes/model/SmartMockHeaders.swift
//
//  SmartMockHeaders.swift
//  SmartMockServer Pod
//
//  Main library model: a set of headers
//

public class SmartMockHeaders {
    
    // --
    // MARK: Members
    // --
    
    var values: [String: [String]] = [:]


    // --
    // MARK: Initialization
    // --
    
    private init() {
        // Private constructor, use factory methods to create an instance
    }
    
    public static func makeFromHeaders(_ headers: [String: [String]]?) -> SmartMockHeaders {
        let result = SmartMockHeaders()
        result.values = headers ?? [:]
        return result
    }
    
    public static func makeFromFlattenedHeaderMap(_ headers: [String: String]) -> SmartMockHeaders {
        let result = SmartMockHeaders()
        for (key, value) in headers {
            let list: [String] = [ value ]
            result.values[key] = list
        }
        return result
    }
    
    
    // --
    // MARK: Access headers
    // --
    
    public func getHeaderMap() -> [String: [String]] {
        return values
    }
    
    public func getFlattenedHeaderMap() -> [String: String] {
        var result: [String: String] = [:]
        for (key, _) in values {
            result[key] = getHeaderValue(key: key)
        }
        return result
    }

    public func overwriteHeaders(_ headers: SmartMockHeaders) {
        for (key, _) in headers.getHeaderMap() {
            setHeader(key: key, value: headers.getHeaderValue(key: key) ?? "")
        }
    }
    
    public func getHeaderValue(key: String) -> String? {
        for (checkKey, checkValue) in values {
            if checkKey.caseInsensitiveCompare(key) == ComparisonResult.orderedSame {
                var result = ""
                for value in checkValue {
                    if result.count > 0 {
                        result += "; "
                    }
                    result += value
                }
                return result
            }
        }
        return nil
    }
    
    public func setHeader(key: String, value: String) {
        let list: [String] = [ value ]
        removeHeader(key: key)
        values[key] = list
    }
    
    public func addHeader(key: String, value: String) {
        for (checkKey, _) in values {
            if checkKey.caseInsensitiveCompare(key) == ComparisonResult.orderedSame {
                values[checkKey]?.append(value)
                return
            }
        }
        setHeader(key: key, value: value)
    }
    
    public func removeHeader(key: String) {
        for (checkKey, _) in values {
            if checkKey.caseInsensitiveCompare(key) == ComparisonResult.orderedSame {
                values.removeValue(forKey: checkKey)
                return
            }
        }
    }
    
}
